hooks/hookstage/invocation.go

```diff
@@ -1,14 +1,89 @@
 package hookstage
 
-import "github.com/prebid/prebid-server/hooks/hookanalytics"
+import (
+	"encoding/json"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/hooks/hookanalytics"
+)
+
+type Entity string
+
+const (
+	EntityHttpRequest              Entity = "http-request"
+	EntityAuctionRequest           Entity = "auction-request"
+	EntityAuctionResponse          Entity = "auction-response"
+	EntityAllProcessedBidResponses Entity = "all-processed-bid-responses"
+)
 
 // InvocationContext holds information passed to module's hook during hook execution.
 type InvocationContext struct {
+	Stage          string
+	AccountId      string
+	Account        *config.Account
+	moduleContexts map[string]ModuleContext
+	mutex          sync.Mutex
+}
+
+func (ctx *InvocationContext) GetModuleContext(moduleCode string) ModuleContext {
+	ctx.mutex.Lock()
+	defer ctx.mutex.Unlock()
```

Comment:
The main reason why mutex was added is because BidderRequest and RawBidderResponse stages will be run in parallel for multiple bidders. Example schema:
```
Bidder appnexus => BidderRequest Stage => Hook Group 1 => Module A
                                                                                                 => Module B
                                                                                                 ...
                                                                    => Hook Group 2 => Module Y
                                                                                                 => Module Z
Bidder avocet      => BidderRequest Stage => Hook Group 1 => Module A
                                                                                                 => Module B
                                                                                                 ...
                                                                    => Hook Group 2 => Module Y
                                                                                                 => Module Z
Other bidders running BidderRequest stage in parallel using same modules ...
```
During these stages we need to protect read/write operations for module context map accessed concurrently by multiple bidders and write operations for stage outcomes slice.